@flexa/fallback state

0xtaf opened the pull request:
# Fallback State

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [x] New Agent
- [ ] Fix an agent error
- [ ] Update an agent logic/test/documentation
- [ ] Code style update (formatting, renaming)
- [ ] Other (please describe): 

## New Agent PR readiness

- [x] `npm install` command works (i.e. package-lock is updated properly)
- [x] The `name` and `description` fields of `package.json` describe the agent properly
- [x] All the libraries listed in `package.json` are being used
- [x] All the development libraries are in `devDependencies` section of `package.json`
- [x] The agent README is updated
- [x] The agent README sufficiently describes what the agent does
- [x] All the alerts that the agent emits are listed in the README
- [ ] The txn/block examples listed in the README generate the expected findings
- [x] Code is easy to follow and/or properly comment
- [x] Agent contains tests
- [x] All tests pass when running `npm tests`
- [x] Tests are not making real network calls (i.e. network interaction is properly mocked)
- [x] Tests cover all the paths (i.e. negative & positive test cases)
- [x] When the tests complete, no warning/errors are shown
- [x] `npm start` works without errors when executed using a jsonRpcUrl for each of the networks specified in the README
- [x] The code do what the README describes
- [x] Network calls are properly being cached (if duplicated calls can occur)
- [x] A performance review has been done (e.g. `Promise.all` is used whenever possible)
- [x] The findings fields are filled out correctly
- [x] Findings `protocol` is specified (if the agent is for an specific one)
- [x] `metadata` of the findings contains all the information needed.
- [x] The SDK methods are being used appropriately (e.g. `filterLog`, `filterFunction`, `getEthersProvider`)
- [x] `ABI`s match the contracts used / `args` used from `LogDescriptions` exists in the `ABI`s 
- [x] the agent follows Forta recommended [best practices](https://docs.forta.network/en/latest/best-practices/)
- [ ] Agent can recover from failed error call when they can fail (e.g. error are being catched on calls than can fail)


## Comments (optional)

The block examples are not listed in the README because I couldn't encounter any.

RCantu92 reviewed (changes requested):
There are commits from the `Withdrawal-Root-Hash-Addition` branch. Please update the commit history.

RCantu92 reviewed (changes requested):
Please add some unit tests for `data.fetcher.ts`.

RCantu92 commented on Flexa-Agents/Fallback-State/README.md:
```diff
@@ -0,0 +1,19 @@
+
+## Description
+
+This agent detects the contract entering a fallback state
```
```suggestion
This agent detects when `FlexaCollateralManager` enters a fallback state
```
Suggestion for a slight re-wording as well as specifying which contract is being referred to.

RCantu92 reviewed (commented):
Overall, good work 👍

e1Ru1o commented on Flexa-Agents/Fallback-State/package.json:
```diff
@@ -0,0 +1,34 @@
+{
+  "name": "fallback-state-agent",
+  "version": "0.0.1",
+  "description": "This agent detects when the contract enters fallback state",
```
```suggestion
  "description": "This agent detects when the Flexa Collateral Manager contract enters fallback state",
```
And update the package-lock accordingly.

e1Ru1o commented on Flexa-Agents/Fallback-State/src/data.fetcher.ts:
```diff
@@ -0,0 +1,32 @@
+  public async getFallbackSetDate(blockNumber: number, flexaAddress: string): Promise<number> {
+    const flexaContract: Contract = new Contract(flexaAddress, flexaInterface, this.provider);
+    const fallbackSetDate: BigNumber = await flexaContract.fallbackSetDate({ blockTag: blockNumber });
+    return BigNumber.from(fallbackSetDate).toNumber();
```
Avoid this conversion. Use `BigNumber` every time you are working with `uint256` because some values will not fit in `number`. Same in the below function.

e1Ru1o commented on Flexa-Agents/Fallback-State/src/agent.ts:
```diff
@@ -0,0 +1,53 @@
+      blockEvent.blockNumber,
+      flexaContractAddress
+    );
+    const previousFallbackSetDate = await fetcher.getFallbackSetDate(blockEvent.blockNumber - 1, flexaContractAddress);
+    const previousFallbackWithdrawalDelaySeconds = await fetcher.getFallbackWithdrawalDelaySeconds(
+      blockEvent.blockNumber - 1,
+      flexaContractAddress
+    );
```
move this calls to inside the `if` in order to only do this queries when it is really needed.

e1Ru1o commented on Flexa-Agents/Fallback-State/src/agent.ts:
```diff
@@ -0,0 +1,53 @@
+            severity: FindingSeverity.Info,
+            type: FindingType.Info,
+            protocol: "Flexa",
+            metadata: {
+              blockTimestamp: blockEvent.block.timestamp.toString(),
+              blockNumber: blockEvent.blockNumber.toString()
+            }
```
I think this metadata is not needed. When you emit the alert, the alert already contains the information relative to the block. At least `blockNumber` would be redundant. Please check the Forta explorer to verify if adding the timestamp is worthy.

0xtaf replied:
you're right. there's also timestamp in forta explorer. I'm removing the metadata.

0xtaf commented on Flexa-Agents/Fallback-State/README.md:
```diff
@@ -0,0 +1,19 @@
+  - Severity is always set to "Info"
+  - Type is always set to "Info"
+  - Metadata fields
+    - blockTimestamp: Timestamp of the block
+    - blockNumber: Block number of the block
```
I'll remove metadata since it's not needed for this agent as Lazaro mentioned below. https://github.com/NethermindEth/Forta-Agents/pull/163#discussion_r821126700

0xtaf commented on Flexa-Agents/Fallback-State/src/agent.spec.ts:
```diff
@@ -0,0 +1,126 @@
+import { TestBlockEvent, createAddress } from "forta-agent-tools";
+import { when, resetAllWhenMocks } from "jest-when";
+
+const factory: string = createAddress("0xa0");
```
since I declared the contract in the constructor, this became redundant. removed it.

0xtaf commented:
Main/important changes:

- Cleared the history from the previous agent.
- Added tests for the fetcher.
- Used contract address as part of the constructor.
- Removed BigNumber > number conversion and updated the related operations.

and resolved everything else you mentioned above.

e1Ru1o commented on Flexa-Agents/Fallback-State/src/data.fetcher.ts:
```diff
@@ -0,0 +1,31 @@
+    return BigNumber.from(fallbackWithdrawalDelaySeconds);
+  }
+
+  public async getPreviousBlockTimestamp(blockNumber: number): Promise<number> {
+    const previousBlockTimestamp = (await this.provider.getBlock(blockNumber - 1)).timestamp;
+    return previousBlockTimestamp;
+  }
```
I think you can change this function to `getBlockTimestamp` and pass `txEvent.blocknumber - 1` in the agent.
The thing is that there is no concept of previous block timestamp, there is only blocks timestamp. The current implementation of the Fetcher contains part agent logic due to this function.

0xtaf commented on Flexa-Agents/Fallback-State/src/data.fetcher.spec.ts:
```diff
@@ -0,0 +1,81 @@
+  });
+
+  it("should return the correct Fallback Withdrawal Delay Seconds", async () => {
+    for (let [flexaAddress, blockNumber, fallbackSetDateCall, fallbackWithdrawalDelaySecondsCall] of TEST_CASES) {
```
it's actually necessary because the parameters are read by their order in this structure: `[flexaAddress, blockNumber, fallbackSetDateCall, fallbackWithdrawalDelaySecondsCall] of TEST_CASES)`.
